pandas/core/series.py

```diff
@@ -3165,7 +3168,12 @@ def _try_cast(arr, take_fast_path):
 
         try:
             subarr = maybe_cast_to_datetime(arr, dtype)
-            if not is_extension_type(subarr):
+            # Take care in creating object arrays (but generators are not
+            # supported, hence the __len__ check):
+            if is_object_dtype(dtype) and (hasattr(subarr, '__len__') and
```

Comment:
aren't you just checking ``is_list_like``?

Reply:
No, for instance
``` python
In [2]: pd.core.dtypes.common.is_list_like((x for x in range(3)))
Out[2]: True
```

I don't know whether there was a general discussion about iterators as input; we could either decide to drop support for them, or to centralize its handling where possible (e.g. at least for indexes and data), in which case I would change these two lines. I can open a new issue for this.

Reply:
ok, then add a function in pandas.core.dtypes.inference to is_generator and use it here (similar to is_iterator)

Reply:
Replacing ``hasattr(subarr, '__len__') `` with ``is_list_like(dtype) and not is_iterator(dtype)`` should do, I don't think we need ``is_generator``. But alternatively, I could add an argument ``is_list_like(iterators=True)``, and use ``is_list_like(iterators=False)`` here. I think it would come handy at several other places.

Reply:
just make a new function, much simpler that way

Reply:
[No need](https://github.com/pandas-dev/pandas/pull/18600#discussion_r167434074)

Reply:
yes there is, we do this elsewhere.

Reply:
Please elaborate on what "new function" would make things "simpler".